File: tools/render.py
"""Render ergogen outline yaml (makerjs models) + points into a labeled SVG in true coordinates.

usage: python render.py OUT_DIR --o name[:color[:width]] ... [--points] [--crop x0,y0,x1,y1] [--scale PX_PER_MM] [--out file.svg] [--grid 5]
"""
import argparse, math, os, sys, yaml

def walk(model, ox=0.0, oy=0.0, acc=None):
    if acc is None:
        acc = []
    if not isinstance(model, dict):
        return acc
    o = model.get('origin') or [0, 0]
    ox2, oy2 = ox + o[0], oy + o[1]
    for p in (model.get('paths') or {}).values():
        po = p.get('origin') or [0, 0]
        px, py = ox2 + po[0], oy2 + po[1]
        t = p.get('type')
        if t == 'line':
            e = p['end']
            acc.append(('line', px, py, ox2 + e[0], oy2 + e[1]))
        elif t == 'circle':
            acc.append(('circle', px, py, p['radius']))
        elif t == 'arc':
            acc.append(('arc', px, py, p['radius'], p['startAngle'], p['endAngle']))
    for m in (model.get('models') or {}).values():
        walk(m, ox2, oy2, acc)
    return acc

def load_outline(out_dir, name):
    with open(os.path.join(out_dir, 'outlines', name + '.yaml'), encoding='utf-8') as f:
        return walk(yaml.safe_load(f))

def bbox(segs):
    xs, ys = [], []
    for s in segs:
        if s[0] == 'line':
            xs += [s[1], s[3]]; ys += [s[2], s[4]]
        elif s[0] in ('circle', 'arc'):
            xs += [s[1] - s[3], s[1] + s[3]]; ys += [s[2] - s[3], s[2] + s[3]]
    return min(xs), min(ys), max(xs), max(ys)

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('out_dir')
    ap.add_argument('--o', action='append', default=[], help='outline[:color[:width]]')
    ap.add_argument('--points', action='store_true')
    ap.add_argument('--labels', action='store_true')
    ap.add_argument('--crop', default=None)
    ap.add_argument('--scale', type=float, default=6.0)
    ap.add_argument('--grid', type=float, default=0)
    ap.add_argument('--out', default='render.svg')
    a = ap.parse_args()

    layers = []
    for spec in a.o:
        parts = spec.split(':')
        name = parts[0]
        color = parts[1] if len(parts) > 1 and parts[1] else '#000'
        width = float(parts[2]) if len(parts) > 2 else 0.2
        layers.append((name, color, width, load_outline(a.out_dir, name)))

    pts = {}
    if a.points or a.labels:
        with open(os.path.join(a.out_dir, 'points', 'points.yaml'), encoding='utf-8') as f:
            pts = yaml.safe_load(f)

    if a.crop:
        x0, y0, x1, y1 = [float(v) for v in a.crop.split(',')]
    else:
        allsegs = [s for l in layers for s in l[3]]
        x0, y0, x1, y1 = bbox(allsegs)
        x0 -= 3; y0 -= 3; x1 += 3; y1 += 3

    S = a.scale
    W, H = (x1 - x0) * S, (y1 - y0) * S
    def X(x): return (x - x0) * S
    def Y(y): return (y1 - y) * S

    out = [f'<svg xmlns="http://www.w3.org/2000/svg" width="{W:.0f}" height="{H:.0f}" viewBox="0 0 {W:.1f} {H:.1f}" style="background:#fff">',
           '<rect width="100%" height="100%" fill="#fff"/>']
    if a.grid:
        g = a.grid
        gx = math.floor(x0 / g) * g
        while gx <= x1:
            out.append(f'<line x1="{X(gx):.1f}" y1="0" x2="{X(gx):.1f}" y2="{H:.1f}" stroke="#eee" stroke-width="1"/>')
            out.append(f'<text x="{X(gx)+2:.1f}" y="10" font-size="9" fill="#999" font-family="monospace">{gx:g}</text>')
            gx += g
        gy = math.floor(y0 / g) * g
        while gy <= y1:
            out.append(f'<line x1="0" y1="{Y(gy):.1f}" x2="{W:.1f}" y2="{Y(gy):.1f}" stroke="#eee" stroke-width="1"/>')
            out.append(f'<text x="2" y="{Y(gy)-2:.1f}" font-size="9" fill="#999" font-family="monospace">{gy:g}</text>')
            gy += g
    for name, color, width, segs in layers:
        out.append(f'<g stroke="{color}" stroke-width="{width*S:.2f}" fill="none" stroke-linecap="round">')
        for s in segs:
            if s[0] == 'line':
                out.append(f'<line x1="{X(s[1]):.2f}" y1="{Y(s[2]):.2f}" x2="{X(s[3]):.2f}" y2="{Y(s[4]):.2f}"/>')
            elif s[0] == 'circle':
                out.append(f'<circle cx="{X(s[1]):.2f}" cy="{Y(s[2]):.2f}" r="{s[3]*S:.2f}"/>')
            elif s[0] == 'arc':
                cx, cy, r, sa, ea = s[1:]
                sweep = (ea - sa) % 360
                if sweep == 0: sweep = 360
                p0 = (cx + r * math.cos(math.radians(sa)), cy + r * math.sin(math.radians(sa)))
                p1 = (cx + r * math.cos(math.radians(ea)), cy + r * math.sin(math.radians(ea)))
                large = 1 if sweep > 180 else 0
                # ccw in math coords -> with flipped y it's cw in svg => sweep-flag 0
                out.append(f'<path d="M {X(p0[0]):.2f} {Y(p0[1]):.2f} A {r*S:.2f} {r*S:.2f} 0 {large} 0 {X(p1[0]):.2f} {Y(p1[1]):.2f}"/>')
        out.append('</g>')
    if pts:
        for n, p in pts.items():
            x, y, r = p['x'], p['y'], p['r']
            if not (x0 <= x <= x1 and y0 <= y <= y1):
                continue
            col = '#d00' if 'key' in (p['meta'].get('tags') or []) else '#06c'
            out.append(f'<circle cx="{X(x):.1f}" cy="{Y(y):.1f}" r="{0.5*S:.1f}" fill="{col}"/>')
            dx, dy = 3 * math.cos(math.radians(r + 90)), 3 * math.sin(math.radians(r + 90))
            out.append(f'<line x1="{X(x):.1f}" y1="{Y(y):.1f}" x2="{X(x+dx):.1f}" y2="{Y(y+dy):.1f}" stroke="{col}" stroke-width="1"/>')
            if a.labels:
                out.append(f'<text x="{X(x)+4:.1f}" y="{Y(y)-3:.1f}" font-size="{max(8, S*1.6):.0f}" fill="{col}" font-family="monospace">{n}</text>')
    out.append('</svg>')
    with open(a.out, 'w', encoding='utf-8') as f:
        f.write('\n'.join(out))
    print(f'wrote {a.out} crop=({x0:.1f},{y0:.1f})-({x1:.1f},{y1:.1f}) size={W:.0f}x{H:.0f}')

if __name__ == '__main__':
    main()

File: tools/render.ts
/**
 * Draws ergogen's outlines and points into a labelled SVG in true coordinates
 * (y up, matching `points.yaml`), so a generated shape can be checked by eye.
 */
import { writeFileSync } from 'node:fs';
import { parseCli, run } from './lib/cli.ts';
import { fixed, general } from './lib/fmt.ts';
import { toRadians, type BBox, type Vec2 } from './lib/geom.ts';
import { bounds, loadOutline, sweepOf, type Outline } from './lib/outline.ts';
import { repoPath } from './lib/paths.ts';
import { hasTag, loadPoints, type Points } from './lib/points.ts';

interface Layer {
  readonly colour: string;
  readonly width: number;
  readonly outline: Outline;
}

/** Maps model coordinates to SVG ones: x shifts, y flips, both scale. */
interface View {
  readonly scale: number;
  readonly width: number;
  readonly height: number;
  x(value: number): number;
  y(value: number): number;
}

const PADDING = 3;
const KEY_COLOUR = '#d00';
const OTHER_COLOUR = '#06c';

function main(): void {
  const args = parseCli({
    usage: 'node tools/render.ts <out> --o name[:colour[:width]] ... [--points] [--labels] [--crop x0,y0,x1,y1] [--scale n] [--grid mm] [--out file.svg]',
    positionals: { dir: { type: 'string', default: 'out', help: 'build output directory' } },
    options: {
      o: { type: 'string', many: true, help: 'outline to draw, repeatable' },
      points: { type: 'boolean', help: 'mark every ergogen point' },
      labels: { type: 'boolean', help: 'name every point' },
      crop: { type: 'string', help: 'x0,y0,x1,y1 in model coordinates' },
      scale: { type: 'number', default: 6, help: 'pixels per mm' },
      grid: { type: 'number', default: 0, help: 'grid spacing in mm, 0 for none' },
      out: { type: 'string', default: 'render.svg', help: 'file to write' },
    },
  });

  const outDir = repoPath(args.dir);
  const layers = args.o.map((spec) => toLayer(outDir, spec));
  const points = args.points || args.labels ? loadPoints(outDir) : {};
  const box = args.crop ? parseCrop(args.crop) : pad(bounds(layers.flatMap((layer) => layer.outline)), PADDING);
  const view = makeView(box, args.scale);

  const svg = [
    `<svg xmlns="http://www.w3.org/2000/svg" width="${fixed(view.width, 0)}" height="${fixed(view.height, 0)}"` +
      ` viewBox="0 0 ${fixed(view.width, 1)} ${fixed(view.height, 1)}" style="background:#fff">`,
    '<rect width="100%" height="100%" fill="#fff"/>',
    ...(args.grid ? grid(box, view, args.grid) : []),
    ...layers.flatMap((layer) => draw(layer, view)),
    ...markPoints(points, box, view, args.labels),
    '</svg>',
  ];
  writeFileSync(repoPath(args.out), svg.join('\n'), 'utf8');

  const extent = `(${fixed(box.minX, 1)},${fixed(box.minY, 1)})-(${fixed(box.maxX, 1)},${fixed(box.maxY, 1)})`;
  console.log(`wrote ${args.out} crop=${extent} size=${fixed(view.width, 0)}x${fixed(view.height, 0)}`);
}

function toLayer(outDir: string, spec: string): Layer {
  const [name = '', colour, width] = spec.split(':');
  return {
    colour: colour || '#000',
    width: width === undefined ? 0.2 : Number(width),
    outline: loadOutline(outDir, name),
  };
}

function makeView(box: BBox, scale: number): View {
  return {
    scale,
    width: (box.maxX - box.minX) * scale,
    height: (box.maxY - box.minY) * scale,
    x: (value) => (value - box.minX) * scale,
    y: (value) => (box.maxY - value) * scale,
  };
}

function grid(box: BBox, view: View, spacing: number): string[] {
  const lines: string[] = [];
  // Accumulating rather than counting, so the last line lands where it used to.
  for (let x = Math.floor(box.minX / spacing) * spacing; x <= box.maxX; x += spacing) {
    lines.push(
      `<line x1="${fixed(view.x(x), 1)}" y1="0" x2="${fixed(view.x(x), 1)}" y2="${fixed(view.height, 1)}" stroke="#eee" stroke-width="1"/>`,
      `<text x="${fixed(view.x(x) + 2, 1)}" y="10" font-size="9" fill="#999" font-family="monospace">${general(x)}</text>`,
    );
  }
  for (let y = Math.floor(box.minY / spacing) * spacing; y <= box.maxY; y += spacing) {
    lines.push(
      `<line x1="0" y1="${fixed(view.y(y), 1)}" x2="${fixed(view.width, 1)}" y2="${fixed(view.y(y), 1)}" stroke="#eee" stroke-width="1"/>`,
      `<text x="2" y="${fixed(view.y(y) - 2, 1)}" font-size="9" fill="#999" font-family="monospace">${general(y)}</text>`,
    );
  }
  return lines;
}

function draw({ colour, width, outline }: Layer, view: View): string[] {
  const shapes = outline.map((segment) => {
    if (segment.kind === 'line') {
      return (
        `<line x1="${fixed(view.x(segment.from[0]), 2)}" y1="${fixed(view.y(segment.from[1]), 2)}"` +
        ` x2="${fixed(view.x(segment.to[0]), 2)}" y2="${fixed(view.y(segment.to[1]), 2)}"/>`
      );
    }
    if (segment.kind === 'circle') {
      return (
        `<circle cx="${fixed(view.x(segment.centre[0]), 2)}" cy="${fixed(view.y(segment.centre[1]), 2)}"` +
        ` r="${fixed(segment.radius * view.scale, 2)}"/>`
      );
    }
    const from = onCircle(segment.centre, segment.radius, segment.start);
    const to = onCircle(segment.centre, segment.radius, segment.end);
    const large = sweepOf(segment) > 180 ? 1 : 0;
    const radius = fixed(segment.radius * view.scale, 2);
    // Counter-clockwise in model coordinates is clockwise once y is flipped,
    // hence the sweep flag of 0.
    return (
      `<path d="M ${fixed(view.x(from[0]), 2)} ${fixed(view.y(from[1]), 2)}` +
      ` A ${radius} ${radius} 0 ${large} 0 ${fixed(view.x(to[0]), 2)} ${fixed(view.y(to[1]), 2)}"/>`
    );
  });
  return [
    `<g stroke="${colour}" stroke-width="${fixed(width * view.scale, 2)}" fill="none" stroke-linecap="round">`,
    ...shapes,
    '</g>',
  ];
}

function markPoints(points: Points, box: BBox, view: View, labels: boolean): string[] {
  const marks: string[] = [];
  for (const [name, point] of Object.entries(points)) {
    const { x, y } = point;
    if (x < box.minX || x > box.maxX || y < box.minY || y > box.maxY) continue;

    const colour = hasTag(point, 'key') ? KEY_COLOUR : OTHER_COLOUR;
    const [dx, dy] = [3 * Math.cos(toRadians(point.r + 90)), 3 * Math.sin(toRadians(point.r + 90))];
    marks.push(
      `<circle cx="${fixed(view.x(x), 1)}" cy="${fixed(view.y(y), 1)}" r="${fixed(0.5 * view.scale, 1)}" fill="${colour}"/>`,
      `<line x1="${fixed(view.x(x), 1)}" y1="${fixed(view.y(y), 1)}"` +
        ` x2="${fixed(view.x(x + dx), 1)}" y2="${fixed(view.y(y + dy), 1)}" stroke="${colour}" stroke-width="1"/>`,
    );
    if (labels) {
      marks.push(
        `<text x="${fixed(view.x(x) + 4, 1)}" y="${fixed(view.y(y) - 3, 1)}"` +
          ` font-size="${fixed(Math.max(8, view.scale * 1.6), 0)}" fill="${colour}" font-family="monospace">${name}</text>`,
      );
    }
  }
  return marks;
}

function parseCrop(spec: string): BBox {
  const [minX, minY, maxX, maxY] = spec.split(',').map(Number);
  if ([minX, minY, maxX, maxY].some((value) => value === undefined || !Number.isFinite(value))) {
    throw new Error(`--crop wants four numbers, got "${spec}"`);
  }
  return { minX: minX!, minY: minY!, maxX: maxX!, maxY: maxY! };
}

const pad = (box: BBox, by: number): BBox => ({
  minX: box.minX - by,
  minY: box.minY - by,
  maxX: box.maxX + by,
  maxY: box.maxY + by,
});

const onCircle = (centre: Vec2, radius: number, degrees: number): Vec2 => [
  centre[0] + radius * Math.cos(toRadians(degrees)),
  centre[1] + radius * Math.sin(toRadians(degrees)),
];

run(main);
